src/shapes/input_text/InputText.js

```diff
@@ -39,40 +56,170 @@ export default class InputText extends Shape {
     let points = []
     let prevLetter = ''
     let x = 0.0
+    let lines = []
     for (let chi = 0; chi < state.shape.inputText.length; chi++) {
-      var letter = state.shape.inputText[chi]
+      var nextLetter = state.shape.inputText[chi]
       if (prevLetter === 'b' || prevLetter === 'v' || prevLetter === 'o' || prevLetter === 'w') {
-        prevLetter = letter
-        if (letter.search('/[a-z]/') === -1 && state.shape.inputFont === 'Cursive')
+        // Save this letter before we possibly add in a '*'
+        prevLetter = nextLetter
+        if (nextLetter.search('[a-z]+') !== -1 && state.shape.inputFont === 'Cursive')
         {
-          letter = letter + '*'
+          nextLetter = nextLetter + '*'
         }
       }
       else {
-        prevLetter = letter
+        prevLetter = nextLetter
+      }
+
+      if (nextLetter === '\n') {
+        // New line
+        lines.push(points)
+        points = []
+        x = 0.0
+        continue
       }
 
       var shape = undefined
       if (state.shape.inputFont === 'Cursive') {
-        shape = CursiveFont(letter)
+        shape = CursiveFont(nextLetter)
       } else if (state.shape.inputFont === 'Sans Serif') {
-        shape = SansSerifFont(letter)
+        shape = SansSerifFont(nextLetter)
       } else if (state.shape.inputFont === 'Monospace') {
-        shape = MonospaceFont(letter)
+        shape = MonospaceFont(nextLetter)
       } else {
         // Internal error, but I'm going to just recover
-        shape = CursiveFont(letter)
+        shape = CursiveFont(nextLetter)
       }
 
+      // TODO add in the "Kern" here.
       for (let vi = 0; vi < shape.vertices.length; vi++) {
         points.push(Vertex(shape.vertices[vi].x + x, shape.vertices[vi].y))
       }
       x += shape.vertices[shape.vertices.length-1].x
     }
-    let widthOffset = x / 2.0
-    return points.map( (point) => {
-      return Vertex(point.x - widthOffset, point.y)
-    })
+    // Save the last line we were working on.
+    lines.push(points)
+
+    // The height of a row of text, including the space above.
+    const maxY = 2.4
+
+    if (state.shape.rotateDir === 'Center') {
+      // Starting Y offset
+      let y = (lines.length - 1) * maxY / 2.0
+      let textPoints = []
+
+      // Capture some wrap around points, to connect the lines.
+      let connectorPoints = []
+      lines.forEach( (points) => {
+        let maxX = getMaxX(points)
+        let widthOffset = maxX / 2.0
+
+        // Add in the connector points (if we have any)
```

Comment:
Not sure this is the relevant line of code, but... can you use traceCircle here? If I recall correctly, it has the added benefit of traveling clockwise or counterclockwise to minimize the distance. I think the text will look better when rotate = "top" or "bottom" if you don't travel most of the connecting circle each time.

Reply:
Great idea.